node/silkworm/stagedsync/stage_execution.cpp

```diff
@@ -109,52 +109,26 @@ StageResult Execution::forward(db::RWTxn& txn) {
     return is_stopping() ? StageResult::kAborted : StageResult::kSuccess;
 }
 
-std::queue<Block> Execution::prefetch_blocks(db::RWTxn& txn, BlockNum from, BlockNum to, size_t max_blocks) {
+void Execution::prefetch_blocks(db::RWTxn& txn, const BlockNum from, const BlockNum to, const size_t max_blocks) {
     std::unique_ptr<StopWatch> sw;
     if (log::test_verbosity(log::Level::kTrace)) {
         sw = std::make_unique<StopWatch>(/*auto_start=*/true);
     }
 
-    std::queue<Block> ret{};
-    auto hashes_table{db::open_cursor(*txn, db::table::kCanonicalHashes)};
-    auto key{db::block_key(from)};
-    auto data{hashes_table.find(db::to_slice(key), true)};
-    while (data.done) {
-        BlockNum reached_block_num{endian::load_big_u64(static_cast<const uint8_t*>(data.key.data()))};
-        if (reached_block_num != from) {
-            throw std::runtime_error("Bad canonical header sequence: expected " + std::to_string(from) + " got " +
-                                     std::to_string(reached_block_num));
-        }
-
-        Bytes block_key(8 + kHashLength, '\0');
-        std::memcpy(&block_key[0], data.key.data(), 8);
-        std::memcpy(&block_key[8], data.value.data(), kHashLength);
-
-        Block block{};
-        auto raw_header{db::read_header_raw(*txn, block_key)};
-        if (raw_header.empty()) {
-            throw std::runtime_error("Unable to load block header for block " + std::to_string(from));
-        }
-        ByteView encoded_header{raw_header.data(), raw_header.length()};
-        rlp::success_or_throw(rlp::decode(encoded_header, block.header));
+    prefetched_blocks_.clear();  // free the memory held by transactions, etc
 
-        if (!db::read_body(*txn, block_key, /*read_senders=*/true, block)) {
-            throw std::runtime_error("Unable to load block body for block " + std::to_string(from));
-        }
-        ret.push(block);
-
-        if (from == to || ret.size() >= max_blocks) {
-            break;
-        }
+    const size_t n{std::min(static_cast<size_t>(to - from + 1), max_blocks)};
+    prefetched_blocks_.resize(n);
```

Comment:
Here we may allocate more blocks than what possibly returned by read_blocks (in case of error). In understand such error would be unrecoverable nevertheless I find this conceptually arguable.